contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.9;
+
+//TODO(security): Replace to in-repo copy of the lib
+import "@openzeppelin/contracts-v4.4/utils/math/SafeCast.sol";
+
+
+/**
+  * @title A dedicated contract for handling stETH withdrawal request queue
+  * @notice it responsible for:
+  * - taking withdrawal requests and placing them in the queue
+  * - finalizing requesuts in queue (making them claimable)
+  * - processing claims for finalized requests
+  * @author folkyatina
+  */
+contract WithdrawalQueue {
+    using SafeCast for uint256;
+    /**
+     * We don't want to deal with small amounts because there is a gas spent on oracle 
+     * for each request. 
+     * But exact threshhold should be defined later when it will be clear how much will 
+     * it cost to withdraw.
+     */
+    uint256 public constant MIN_WITHDRAWAL = 0.1 ether;
+
+    /**
+     * @notice All state-modifying calls are allowed only from owner protocol. 
+     * @dev should be Lido
+     */ 
+    address public immutable OWNER;
+
+    /**
+     * @notice amount of ETH on this contract balance that is locked for withdrawal and waiting for claim
+     * @dev Invariant: `lockedEtherAmount <= this.balance`
+     */
+    uint256 public lockedEtherAmount = 0;
+
+    /**
+     * @notice queue for withdrawal requests
+     */ 
+    mapping(uint => Request) public queue;
+    
+    uint256 public queueLength = 0;
+    uint256 public finalizedQueueLength = 0;
```

Comment:
lastFinalizedRequestId does not fit well, because it can't represent the state with no finalized requests at all in a clear manner. So, it better to stick with the current way of doing it, but we can think about renaming.